File: 133. Clone Graph/solution.py
# Definition for a undirected graph node
# class UndirectedGraphNode:
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []

class Solution:
    # @param node, a undirected graph node
    # @return a undirected graph node
    def __init__(self):
        self.graph = {}
    
    def cloneGraph(self, node):
        if not node:
            return None
        
        if node in self.graph:
            return self.graph[node]
        
        clone = UndirectedGraphNode(node.label)
        self.graph[node] = clone
        
        for neigh in node.neighbors:
            clone.neighbors.append(self.cloneGraph(neigh))
            
        return clone
            

File: 133. Clone Graph/solution.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;

    public Node() {}

    public Node(int _val,List<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/
class Solution {
    public HashMap<Node, Node> visited = new HashMap<Node, Node>();
    public Node cloneGraph(Node node) {
        if (node == null) {
            return null;
        }
        
        Node clone = new Node(node.val);
        clone.neighbors = new ArrayList<Node>();
        visited.put(node, clone);
        for (Node neigh: node.neighbors) {
            if (neigh != null) {
                if (visited.containsKey(neigh)) {
                    clone.neighbors.add(visited.get(neigh));
                } else {
                    clone.neighbors.add(cloneGraph(neigh));
                }
            }
        }
        return clone;
    }
}
